Perfmageddon

disconcision opened the pull request:
Currently offers an up to an order-of-magnitude speedup on best-case edit action keystrokes

The biggest gains are due to eliminating extraneous Incr_dom cycles happening on ~ every keystroke:

- Due to the legacy doubletap detection, every KeyUp event was triggering a new cycle. Removing this gives a ~50%+ overall speedup
- Saving was happening every keystroke, in fact sometimes twice (on keyup as well). This was a double-whammy, as saving on edit actions was being done in Keyboard by scheduling an additional event. The ability of the key handlers to return a list of updates, each of which triggers a full incr_dom cycle, is a moral hazard and has been removed; each key event now returns an option(Action) instead. Saving also no longer happens after every edit action; rather it happens 1 second after the last edit action happened. Together this is a ~200-400% overall speedup.
- There was a check in the on_display hook which scrolled the caret into view which was gated on detecting a change in the model. This is expensive, not because of the model comparison, but simply because the old_model is in the Incr monad, and mapping into Incr triggers a cycle. Removing this check gives a ~50% speedup.

Other gains:

- 5x speedup to Paste by bypassing Editor Meta recalculation
- Huge speedup in very imbalanced states by memoizing Backpack.shard_info
- 10x speedup in Holes Deco by just scanning info_map instead of recursing
- 2x speedup for Termranges.mk; was called twice with no memoization. I now passing it through directly as it's clean to do so, but I also memoized it defensively
- 60% speedup for Code.view by recursively memoizing Text.of_tile
- 30% speedup for MakeTerm by memoizing Skel.mk

Remaining possible speedups:

- Logging: Gets slow with lots of entries. With localstore you can't just append, you need to set. Could switch to a system for which each entry gets its own a KV pair of something.
- Saving Editors to LocalStorage: I caused a bit of a regression when I moved to serializing the entire Editors. Probably not a priority with the new save delay system, but could use some TLC eventually.
- Backpack: If we are still experiencing too much slowdown in unbalanced states, shard_info could be greatly speeded up if we were to eliminate unrestricted pick-up functionality (i.e. the ability to have backpack selections containing multiple shards)
- Maketerm: The recursion structure in Maketerm makes recursive memoization difficult. We want to memoize something that's likely to have physical equality i.e. substructures which are likely literally preserved over edits, like segments or tiles. But the core Maketerm functions (go_s, unsorted) take both a segment and a skel as arguments; since the skel is derived, we're never going to have physical equality for substructures if the superstructure changes. I'm wondering if it's possible to internalize skel generation to these functions; instead of sending in a skel, send in a subsegment
- PtMaps/Maps/Hashtbls: Replace remaining ptmap use with plain maps and/or replace maps use with hashtbls everywhere. I'm unsure as to the overall impact but it could be significant; probably worth doing more bottom-up profiling. As a stopgap we're still unioning in some places which could be turned into traversals (e.g. TermRanges)
- Elaboration Views: Right now turning off dynamics can slow things down as it takes time to render large elaborations. This is partially addressed in `llmass` as it makes it possible to actually turn of statics, resulting in no elaboration being generated. There should probably be a separate toggle for elaboration though.
- Zipper pass fusion: In general we could probably consider some more structured architecture for passes over the zipper. Currently we have MakeTerm, Measured, TermRanges, and Code.view which all do zipper passes. Two or more of these could potentially be fused into a single 'SyntaxInfo' pass. SyntaxInfo could potentially collect some information which is either being reproduced ad-hoc in many places (expected sort) or currently lives in static info (term, cls, list of ancestor ids)

disconcision committed:
perf

disconcision committed:
instead of saving immediately after edits, wait until 2 seconds after last edit

disconcision committed:
change save wait time to 1 second

disconcision committed:
memoize Text.of_tile and Segment.skel

disconcision committed:
memoize shard_info for large perf improvement when there are lots of things in the backpack

disconcision committed:
5x speedup to paste

disconcision committed:
zipper_of_string cleanup

disconcision committed:
Merge branch 'randIds' of https://github.com/hazelgrove/hazel into perfmageddon

disconcision committed:
rec memo termranges

disconcision committed:
merge fix?

disconcision committed:
paste linebreak fix

disconcision committed:
paste linebreak fix 2

disconcision committed:
paste linebreak fix 3

disconcision committed:
Merge branch 'randIds' of https://github.com/hazelgrove/hazel into perfmageddon

disconcision committed:
Merge branch 'randIds' of https://github.com/hazelgrove/hazel into perfmageddon

disconcision committed:
fix memo bug in Code manifesting in code text and deco being offset

cyrus- committed:
Merge branch 'dev' into perfmageddon

cyrus- reviewed (changes requested):
getting `Not_found` exceptions when editing some school mode editors, e.g. the Prelude. Maybe having to do with serialization issues?

disconcision committed:
fix Not_found exn

disconcision committed:
Merge branch 'perfmageddon' of https://github.com/hazelgrove/hazel into perfmageddon

disconcision commented:
@cyrus- fixed. the issue wasn't specific to exercises mode, it was something I had noticed and fixed a while ago, the fix just ended up on the wrong side of the permageddon/llmass fix. I realize the comment I've added to the fix isn't totally satisfying; it's a complex/finicky syntax issue that predates this branch; it is only surfaced here because previously we've never been iterating over the info_map directly, only recursing though the AST

disconcision committed:
Merge branch 'dev' into perfmageddon
